Controller/Gateway/MoneroPayment.php

```diff
@@ -330,63 +336,129 @@ public function retriveprice($currency)
                 return $price;
         }
     }
-    
-    public function paymentid_cookie()
+
+    public function subaddress_cookie()
     {
-        if (!isset($_COOKIE['payment_id']))
-        {
-            $payment_id = bin2hex(openssl_random_pseudo_bytes(8));
-            setcookie('payment_id', $payment_id, time() + 2700);
+        if (!isset($_COOKIE['xmr_subaddress'])) {
+            // Generate subaddress
+            $xmr_subaddress = $this->monero_daemon->create_address(0);
+            setcookie('xmr_subaddress', $xmr_subaddress['address'], time() + 2700);
+        } else {
+            $xmr_subaddress = ['address' => $_COOKIE['xmr_subaddress']];
         }
-        else
-            $payment_id = $_COOKIE['payment_id'];
-        return $payment_id;
+        return $xmr_subaddress['address'];
     }
-    
+
     public function changeto($amount, $currency)
     {
         $rate = $this->retriveprice($currency);
         $price_converted = $amount / $rate;
         $converted_rounded = round($price_converted, 11); //the Monero wallet can't handle decimals smaller than 0.000000000001
-            return $converted_rounded;
+        return $converted_rounded;
+    }
+
+    protected function check_payment_rpc($subaddress)
+    {
+        $txs = array();
+        $address_index = $this->monero_daemon->get_address_index($subaddress);
+        if(isset($address_index['index']['minor'])){
+          $address_index = $address_index['index']['minor'];
+        }
+        else {
+          return $txs;
```

Comment:
there's something odd here..

Write
```
if(!isset(...)){
    return $txs;
}
```
and move the other branch after that check :) better readability

Reply:
done thnx